### ISSUE TITLE
.Net Core 3.1 problem when Hosting on IIS

### ISSUE THREAD

**Issue by AminEsmaeily** [dotnet/aspnetcore#18481]
I have an ASP.Net Core project that I started it using the .Net Core 2.2. I hosted it on the IIS 10 and windows server 2016 and it was working perfectly.  I upgraded the entire solution to 3.0 and then 3.1 after a while. And hosted it on the IIS. My current server is Windows Server 2019 and the IIS version is 10.0.17763.1, but when I browse the website address using the web browser it shows the 500.19(0x8007000d) error. I have installed the version 3.1.0 of the .Net Core Hosting Bundle on the system and following runtimes are installed with it:
![image](https://user-images.githubusercontent.com/11803924/72662985-aae98e00-3a02-11ea-8fce-388d51febd33.png)

After lots of tests, I found that the problem goes out by installing the version 3.0.1 of the .Net Core Hosting Bundle. This package adds the following runtimes to the system:
![image](https://user-images.githubusercontent.com/11803924/72663041-395e0f80-3a03-11ea-9671-93a1f032358d.png)
It seems that the `Microsoft.AspNetCore.App 3.0.1` is needed or covers the deficiency of the `Microsoft.AspNetCore.App 3.1.0`.

**Comment by shirhatti** [comment]
500.19 indicates a malformed IIS configuration. A common reason why folks run into this is because they end up having a `web.config` that references the `aspNetCoreModule` without having it installed.

My hypothesis here is that when the 3.1.0 bundle was installed, IIS was not enabled on the machine and ergo the IIS component of the bundle was skipped. When you later installed the 3.0.1 bundle, IIS happened to be enabled and hence the IIS component of the bundle was installed.

One way to validate my hypothesis is to uninstall the 3.0.1 bundle followed by a repair of the 3.1.0 bundle. My guess is your application will continue to work after that. If it doesn't, then my hypothesis is incorrect and we'll need to investigate further.

@AminEsmaeily Could attempt to perform my suggested steps?

**Comment by AminEsmaeily** [comment]
Dear @shirhatti, to answer your hypothesis and check all of the steps, I start the hosting process from scratch. To test the whole process, I use a VM with **Windows 10 Enterprise Build 14393 RS1** and the **IIS version 10.0.14393.0**. My target ASP.Net core project is based on **.Net Core 3.1** that comes with multiple **.Net Standard 2.1** class libraries. The _web.config_ file is as follows:
``` xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <system.webServer>
    <security>
      <requestFiltering>
        <requestLimits maxAllowedContentLength="2147483648"></requestLimits>
      </requestFiltering>
    </security>
  </system.webServer>
  <location path="." inheritInChildApplications="false">
    <system.webServer>
      <aspNetCore processPath="%LAUNCHER_PATH%" arguments="%LAUNCHER_ARGS%" stdoutLogEnabled="false" hostingModel="InProcess">
      </aspNetCore>
      <handlers>
        <add name="aspNetCore" path="*" verb="*" modules="AspNetCoreModuleV2" resourceType="Unspecified" />
      </handlers>
    </system.webServer>
  </location>
</configuration>
```
I checked the environment to ensure that nothing about the **.Net Core** exists on the system
![image](https://user-images.githubusercontent.com/11803924/73116890-3effac00-3f53-11ea-815e-927eb7bb89e2.png)
And my web application on the IIS shows the current page:
![image](https://user-images.githubusercontent.com/11803924/73116913-8be38280-3f53-11ea-8c41-86d2a889f314.png)
The IIS App Pool is configured as follows:
![image](https://user-images.githubusercontent.com/11803924/73116927-cbaa6a00-3f53-11ea-859a-0333137f2168.png)
For the first step, I install the **.Net Core Hosting Bundle 3.0.1 - x64**
![image](https://user-images.githubusercontent.com/11803924/73116997-95b9b580-3f54-11ea-9d33-0856a373cf7d.png)
When I run it on the IIS, It shows the below page:
![image](https://user-images.githubusercontent.com/11803924/73117080-bfbfa780-3f55-11ea-8306-8c22ce232253.png)
Then I checked it using the `dotnet` command and it showed me the current message:
![image](https://user-images.githubusercontent.com/11803924/73117101-eaa9fb80-3f55-11ea-9bc2-fa00bc95466f.png)
To fix this problem, I made some changes in the `aspNetCore` node of the **web.config**:
``` xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <system.webServer>
    <security>
      <requestFiltering>
        <requestLimits maxAllowedContentLength="2147483648"></requestLimits>
      </requestFiltering>
    </security>
  </system.webServer>
  <location path="." inheritInChildApplications="false">
    <system.webServer>
      <aspNetCore processPath="dotnet" arguments=".\LifeStyle.dll" stdoutLogEnabled="false" hostingModel="InProcess" stdoutLogFile=".\logs\stdout">
      </aspNetCore>
      <handlers>
        <add name="aspNetCore" path="*" verb="*" modules="AspNetCoreModuleV2" resourceType="Unspecified" />
      </handlers>
    </system.webServer>
  </location>
</configuration>
```
By applying these changes, the message changed to the following:
![image](https://user-images.githubusercontent.com/11803924/73117433-d4526e80-3f5a-11ea-83c9-b0537af8db1c.png)
For the second step, I removed the **Hosting bundle 3.0.1** and installed the version **3.1.0**. By doing this, my application came up running:
![image](https://user-images.githubusercontent.com/11803924/73117521-0912f580-3f5c-11ea-9e9c-d7bf9b3d83d4.png)

It seems that changing the **LAUNCHER_PATH** and **LAUNCHER_ARGS** to the explicit values fixed the problem.

**Comment by anurse** [comment]
@AminEsmaeily the root cause of the problem sounds like what @shirhatti said. The 500.19 error occurs because you installed the hosting bundle before IIS. The testing you did later seems to indicate expected behavior. The `LAUNCHER_PATH` and `LAUNCHER_ARGS` is not what fixed the problem, it was installing the 3.1.0 Hosting Bundle. You indicated your app targets 3.1.0 which means that until you install the 3.1.0 runtime (included in the Hosting Bundle) your app will fail with 500.31.

It sounds like you've gotten your app up and running, so I don't see actionable work for us here. Closing.
